pkg/kapis/gitops/v1alpha1/handler.go

```diff
@@ -0,0 +1,154 @@
+	healthStatus := common.GetQueryParameter(req, healthStatusQueryParam)
+	syncStatus := common.GetQueryParameter(req, syncStatusQueryParam)
+
+	applicationList := &v1alpha1.ApplicationList{}
+	matchingLabels := client.MatchingLabels{}
+	if syncStatus != "" {
+		matchingLabels[v1alpha1.SyncStatusLabelKey] = syncStatus
+	}
+	if healthStatus != "" {
+		matchingLabels[v1alpha1.HealthStatusLabelKey] = healthStatus
+	}
+	if err := h.List(context.Background(), applicationList, client.InNamespace(namespace), matchingLabels); err != nil {
+		common.Response(req, res, applicationList, err)
+		return
+	}
+
+	queryParam := query.ParseQueryParameter(req)
+	list := v1alpha3.DefaultList(toObjects(applicationList.Items), queryParam, v1alpha3.DefaultCompare(), v1alpha3.DefaultFilter(), nil)
+
+	common.Response(req, res, list, nil)
+}
+
+func (h *handler) getApplication(req *restful.Request, res *restful.Response) {
+	namespace := common.GetPathParameter(req, common.NamespacePathParameter)
+	name := common.GetPathParameter(req, pathParameterApplication)
+
+	application := &v1alpha1.Application{}
+	err := h.Get(context.Background(), types.NamespacedName{
+		Namespace: namespace,
+		Name:      name,
+	}, application)
+	common.Response(req, res, application, err)
+}
+
+func (h *handler) delApplication(req *restful.Request, res *restful.Response) {
+	namespace := common.GetPathParameter(req, common.NamespacePathParameter)
+	name := common.GetPathParameter(req, pathParameterApplication)
+	cascade := common.GetQueryParameter(req, cascadeQueryParam)
+
+	ctx := context.Background()
+	application := &v1alpha1.Application{}
+	objectKey := types.NamespacedName{
+		Namespace: namespace,
+		Name:      name,
+	}
+	err := h.Get(ctx, objectKey, application)
+	if err == nil {
+		switch application.Spec.Kind {
+		case v1alpha1.ArgoCD:
+			// add the Argo CD resources finalizer if cascade is true
+			if cascade == "true" {
+				if k8sutil.AddFinalizer(&application.ObjectMeta, v1alpha1.ArgoCDResourcesFinalizer) {
+					if err = h.Update(ctx, application); err != nil {
+						common.Response(req, res, application, err)
+						return
+					}
+
+					if err = h.Get(ctx, objectKey, application); err != nil {
+						common.Response(req, res, application, err)
+						return
+					}
+				}
+			}
+		case v1alpha1.FluxCD:
+		}
+		err = h.Delete(ctx, application)
+	}
+	common.Response(req, res, application, err)
+}
+
+func (h *handler) createApplication(req *restful.Request, res *restful.Response) {
+	var err error
+	namespace := common.GetPathParameter(req, common.NamespacePathParameter)
+
+	application := &v1alpha1.Application{}
+	if err = req.ReadEntity(application); err == nil {
+		application.Namespace = namespace
+		switch application.Spec.Kind {
+		case v1alpha1.ArgoCD:
+			if application.Labels == nil {
+				application.Labels = make(map[string]string)
+			}
+			application.Labels[v1alpha1.ArgoCDLocationLabelKey] = h.ArgoCDNamespace
```

Comment:
@lxm Yes. you're right. the public handler shouldn't use argoOption theoretically. But it need this ArgoCDNamespace to create a argoApp. We can actually split this createApplication func into argo and flux handler. I don't which is the best way actually.